pkg/kv/kvserver/batcheval/cmd_add_sstable.go

```diff
@@ -248,8 +259,10 @@ func EvalAddSSTable(
 	return result.Result{
 		Replicated: kvserverpb.ReplicatedEvalResult{
 			AddSSTable: &kvserverpb.ReplicatedEvalResult_AddSSTable{
-				Data:  sst,
-				CRC32: util.CRC32(sst),
+				Data:          sst,
+				CRC32:         util.CRC32(sst),
+				Span:          roachpb.Span{Key: start.Key, EndKey: end.Key},
+				EmitRangefeed: args.WriteAtRequestTimestamp,
```

Comment:
I wonder if it'd make sense to persist this bool as "AtWriteTimestamp" or something like that, and then below instead of skipping non-mvcc AddSSTables, go ahead and send them to the feed until we get to `ev.SST != nil` and then there, if the SST was _not_ AtWriteTimestamp, fail the feed because we got an unexpected violation of the mvcc semantics that feed is relying on?

Reply:
Yeah, I think that makes sense -- describe the SST itself, and move the rangefeed policy elsewhere. I'll update the PR.

Reply:
I think it might make sense to call this `AtRequestTimestamp` and defer the rangefeed policy elsewhere, such that the data describes the SST rather than prescribing downstream actions. Even though it may be a bit unfortunate to bury this policy deep in the bowels of Raft state machine application.

However, given the size of these SSTs, I'd rather not ship all of them through the rangefeed machinery unnecessarily. Also, we only propagate the write timestamp into the rangefeed, with the assumption that all MVCC timestamps in the SST are equal to it:

https://github.com/cockroachdb/cockroach/blob/b3a84a243eeb2c98455a17e51b79b17e9ac97ef1/pkg/kv/kvserver/replica_rangefeed.go#L617-L619

It seems unnecessary to add a boolean parameter here, and then assert that the passed parameter is always `true`. Shouldn't it be the caller's responsibility to only call the method when appropriate?

Reply:
Then again, maybe we could move the policy into `Processor.ConsumeSSTable` such that it ignores `AtWriteTimestamp == false`

Reply:
My initial suggestion would be to move it further beyond raft, all the way to `*RangeFeed.processEvents`, where it would be an _error_ if an sst event arrived that was not marked as `AtRequestTimestamp`. If we were just ignoring them at that point I'd be more concerned about the cost of shipping them all the way through just to discard them, but I was imaging that if we kick out an assertion failure error or something when the very first arrives that tears down the feed then that cost isn't a big deal since it quickly, and importantly loudly, comes to a stop?

Reply:
Ok, but I think I'd prefer to assert that on the server-side at least, e.g. inside the rangefeed processor at `Processor.publishSSTable`, rather than waiting for it to make its way out to the client and assert it there. In the latter case, we'd have to implement this in all clients (not just the `Factory`-constructed ones which aren't always used), along with the necessary plumbing.

Reply:
Should we maybe do this for all other non-MVCC operations such `ClearRange` and `RevertRange` too? I.e. signal to the rangefeed that we applied a non-MVCC operation, and disconnect it?

Reply:
That sounds good to me.

Reply:
Done, see separate commit which disconnects rangefeeds with a permanent `MVCCHistoryMutationError` for operations that mutate MVCC history -- specifically `AddSSTable` when not writing at the request timestamp, `ClearRange`, and `RevertRange`. This does *not* apply to MVCC-compliant `AddSSTable` requests that write at the request timestamp, which are emitted across rangefeeds.

Note that for `rangefeed.Factory`-based clients, this is surfaced via `OnInternalError` due to the current API design -- unfortunately, most callers do not set this callback, and will silently ignore the error even though the rangefeed stops. The owning teams should add appropriate error handling here.